### ISSUE TITLE
Create offline source package

### ISSUE THREAD

**Issue by xylosper** [conan-io/conan#4609]
Maybe #2197 related.

I need a functionality to create offline source package.
Generally, a conanfile requires some remote source archives.
In my workspace, I have to work in intranet disconnected from internet connection for security issue.
For now, I'm making all packages contain offline source code using `exports_sources` attribute instead `source()` function one by one by myself.
It would be nice to have automatic generation of offline source package from already existing conanfile.
Just compiling package from outside of intranet and copying it won't be sufficient because each developer may works in different environment and it's too much job to make binaries for various environment combination.

**Comment by Johnnyxy** [comment]
You could setup your own Conan server and sync all your Conan related stuff only with this server.

1. setup Conan server
2. download source (like boost, etc.) onto a intranet FTP/HTTP/Git server
3. create or modify recipes to download the source from the intranet server (see [FTP](https://docs.conan.io/en/latest/reference/tools.html#tools-ftp-download)/[HTTP](https://docs.conan.io/en/latest/reference/tools.html#tools-download)/[Git](https://docs.conan.io/en/latest/reference/tools.html#tools-git))
4. build your recipes only with depending only on references ([localuser/localchannel](https://docs.conan.io/en/latest/introduction.html#binary-management)) from your local builds (do not use conan/stable or something official as you could not distinguish between your builds and [conan-center](https://bintray.com/conan/conan-center)/bintray-builds.
5. consumers use only your own package references

This is some work but not that hard or complicated. One disadvantage is that you cannot sync your recipes with official ones (like from conan-center) automatically. You always have to do this manually and upload them to your local conan server. But normally in larger projects you do not switch to a new version of a dependency every day.

**Comment by memsharded** [comment]
Hi @xylosper 

Let me understand your use case.

Lets say you have a recipe like:

```python
class Pkg(ConanFile):
     def source(self):
         tools.download("http://some.external.url/sources.zip")
```

And the problem is that in your intranet you don't have access to that external url ``"http://some.external.url/sources.zip"``

Quick question, why don't you download the ``sources.zip`` and copy it to a generic Artifactory repo, and then change the URL of the recipe? I'd say that is the way I'd do it.

**Comment by uilianries** [comment]
Also, you can download the zip file and export it with Conan.

```python
class PkgConan(self):
    exports_sources = "foobar.zip"
    def source(self):
        tools.unzip("foobar.zip")        
```

**Comment by milahu** [comment]
> without editing the conanfile

https caching proxy ... ? just need to fix the ssl certificates

**Comment by TsXor** [comment]
I suddenly came up with an idea to create "offline" source packages like python `sdist`.

According to [document on local development flow](https://docs.conan.io/2/tutorial/developing_packages/local_package_development_flow.html), after executing `conan source`, the remote source needed will be put into `src` folder. Then, we can use `conan install` to gather dependencies and use `conan build` to build the package for different profiles. Finally, we can use `conan export-pkg` to export binaries to local cache.
Also, there's a `conan export` command to export recipe to local cache.

Putting these together, we can run `conan source` on a recipe and simply zip the recipe folder with downloaded source and distribute the archive. Users of these archives just execute `conan install`, `conan export-pkg` and `conan export`, then they get recipes and binaries in their local cache.

**Comment by memsharded** [comment]
This is a very old ticket, many things have changed since. 
Conan now contains new features, including for example the "sources backup": https://blog.conan.io/2023/10/03/backup-sources-feature.html, that allows to do a copy of internet sources into your own repo, for further reproducibility. This is for third-party code. 

Conan recipes already do a copy and redistribution of sources if necessary (export_sources attribute or method):

For first party code, just git repos with a ``conanfile`` in it, no need to distribute source packages.

**Comment by TsXor** [comment]
> including for example the "sources backup"

I gave up on this when I saw "To use this feature, the first step is to have a file server. "

I know conan have paid lots of effort in implementing equivalent features. We have deploy and sources backup (I've done search and read document about them before posting the comment above), and yes, they meet the requirements.
But I don't understand why everything need a server. I need a server to make backups. I need a server to make a custom repository. Why can't I simply bring a simple fix by giving an archive? Why can't I say "try using this patched sdist and it's done" like in python?

I don't know. Maybe it's because I'm a silly person that have never attend a job on devops.

I know words above (and following, of course) are very very naive. But conan seems to just let me choose one of the following:
- Buy a web server to distribute your custom packages.
- Acquire access to ConanCenter then submit your package.
- Still buy a web server to back up sources.

These ways depend too much on what is irrelevant to solving the problem "why this package cannot compile".

**Comment by memsharded** [comment]
> But I don't understand why everything need a server. I need a server to make backups. I need a server to make a custom repository. Why can't I simply bring a simple fix by giving an archive? Why can't I say "try using this patched sdist and it's done" like in python?

You can also use the local-recipes-index feature to work at the recipe-source level as conan-center-index does: https://blog.conan.io/2024/04/23/Introducing-local-recipes-index-remote.html

**Comment by TsXor** [comment]
> You can also use the local-recipes-index feature

If I didn't get it wrong, this feature is designed for testing changes to ConanCenter.
Also, you will have to clone an entire recipe repository just for using any of them.

I have to clarify that I don't mean to try to alter conan's design by my words. I sincerely apologize for any trouble caused.

Still, I just want to say that conan's design is just "custom centralized". It is hard to live without a center, but users can build a center themselves.

Conan developers seems to be keeping an idea that "we don't need a portable package file and it is just fine to transfer anything from local cache to remote cache and then to another local cache" (yes, I know we have `conan cache save`). Well, then why will tarballs, sdists and deb/rpm/... packages ever exist? Are they all meaningless?

**Comment by memsharded** [comment]
> If I didn't get it wrong, this feature is designed for testing changes to ConanCenter.

No, not really. You can use it for your own packages that don't exist in ConanCenter. You can have a repo with just some recipes, not related to ConanCenter. As long as the layout is respected, it can work.

> Also, you will have to clone an entire recipe repository just for using any of them.

Yes, but this is a very cheap operation, specially compared with having to build from sources because not using a server that will store the binaries

>  I have to clarify that I don't mean to try to alter conan's design by my words. I sincerely apologize for any trouble caused.

No trouble at all! 🙂 
We really appreciate all feedback, and we are always happy to help. If it doesn't sound like that, it is because sometimes we are a bit overwhelmed and we just rush over tickets and other things. Don't hesitate to create tickets for any question or thing you want to discuss. Sometimes we will not be able to follow up as much as we want, but it is always better to have the feedback that missing it.

> Still, I just want to say that conan's design is just "custom centralized". It is hard to live without a center, but users can build a center themselves.

This is true, Conan is optimized for the "organization" or more enterprise usage. Many of its core features as the scalable binary management are intended for having fast, reproducible builds, but that also means that dependency resolution cannot be easily implemented without a proper server. For example, take into account that Conan can install multiple versions of the same package, and for each version multiple different binaries, with the same ``conan install`` without conflicting, so this model cannot simply match a "tarball" origin.

> Conan developers seems to be keeping an idea that "we don't need a portable package file and it is just fine to transfer anything from local cache to remote cache and then to another local cache" (yes, I know we have conan cache save). Well, then why will tarballs, sdists and deb/rpm/... packages ever exist? Are they all meaningless?

Yes, exactly, packages can also be moved around with ``conan cache save``.

The ``deb/rpm`` packages are also similar to Python ones, they follow the same paradigm of defining and using a set of packages that work well together for a given distro, and things are guaranteed at the distro level. But language package managers are very different from system package managers. A better comparison would be comparing it with Maven for Java, which is very similar to the Conan model (Conan inspired in their model) that always use a server, with a local shared cache for installed packages.
